src/lib/storage/disk_analyzer.rb

```diff
@@ -0,0 +1,368 @@
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "fileutils"
+require "pp"
+
+module Yast
+  module Storage
+    #
+    # Class to analyze the disks (the storage setup) of the existing system:
+    # Check the existing disks and their partitions what candidates there are
+    # to install on, typically eliminate the installation media from that list
+    # (unless there is no other disk), check if there already are any
+    # partitions that look like there was a Linux system previously installed
+    # on that machine, check if there is a Windows partition that could be
+    # resized.
+    #
+    # Some of those operations involve trying to mount the underlying
+    # filesystem.
+    #
+    class DiskAnalyzer
+      include Yast::Logger
+
+      LINUX_PARTITION_IDS =
+        [
+          ::Storage::ID_LINUX,
+          ::Storage::ID_SWAP,
+          ::Storage::ID_LVM,
+          ::Storage::ID_RAID
+        ]
+
+      WINDOWS_PARTITION_IDS =
+        [
+          ::Storage::ID_NTFS,
+          ::Storage::ID_DOS32,
+          ::Storage::ID_DOS16,
+          ::Storage::ID_DOS12
+        ]
+
+      DEFAULT_DISK_CHECK_LIMIT = 10
+
+      attr_reader :installation_disks, :candidate_disks
+      attr_reader :windows_partitions, :linux_partitions
+      attr_accessor :disk_check_limit
+
+      def initialize
+        Yast.import "Arch"
+
+        @installation_disks = []
+        @candidate_disks    = []
+        @linux_partitions   = []
+        @windows_partitions = [] # only filled if @linux_partitions is empty
+
+        # Maximum number of disks to check. This might be important on
+        # architectures that tend to have very many disks (s390).
+        @disk_check_limit = DEFAULT_DISK_CHECK_LIMIT
+      end
+
+      # Analyze disks and partitions. Make sure to call this before querying
+      # any member variables.
+      #
+      def analyze
+        @installation_disks = find_installation_disks
+        @candidate_disks    = find_candidate_disks
+        @linux_partitions   = find_linux_partitions
+
+        if @linux_partitions.empty?
+          @windows_partitions = find_windows_partitions
+        else
+          # We only want to resize any MS Windows partition if there is no
+          # Linux already on that machine, so we don't need to check for
+          # Windows partitions in that case. This saves mounting and unmounting
+          # all Windows-like partitions.
+          log.info("Linux partitions found - not checking for Windows partitions")
+        end
+
+        log.info("Installation disks: #{dev_names(@installation_disks)}")
+        log.info("Candidate    disks: #{dev_names(@candidate_disks)}")
+        log.info("Linux   partitions: #{dev_names(@linux_partitions)}")
+        log.info("Windows partitions: #{dev_names(@windows_partitions)}")
+      end
```

Comment:
I expect here access limitation like `private` as there is only few public methods. It really helps e.g. in generated documentation if you have only two public methods instead of ten.